compiler/backend/src/org/jetbrains/kotlin/codegen/coroutines/coroutines-codegen.md

```diff
@@ -0,0 +1,3266 @@
+Let us now have a look into how we deal with suspend function parameters. We do not generate fields for them, since a lambda uses them just
+to pass the arguments from `invoke` to `invokeSuspend`. We do not need them for suspending functions: the arguments are locals; thus, we
+reuse local variables spilling for them.
+
+Nevertheless, keeping the parameters in the function signature breaks resumption. There is not enough information in continuation's
+`invokeSuspend` to pass them as they were before or as they are now. So, we just put nulls for reference types and zeroes for primitives.
+That means we cannot generate nullability checks at the beginning of the function, so they must be generated at the beginning of the first
+state, to which we cannot resume.
+
+For example, if we change the `test` function to accept an argument:
+```kotlin
+suspend fun dummy() {}
+
+suspend fun test(a: Int) {
+    dummy()
+    dummy()
+}
+```
+its continuation's `invokeSuspend` becomes something like
+```kotlin
+fun invokeSuspend(result: Result<Any?>): Any? {
+    this.result = result
+    this.label.sign_bit = 1
+    test(0, this)
+}
+```
+
+### JVM: Layout
+
+We can now deduce the layout of the suspend function's continuation.
+
+The ideal suspend lambda layout is the following:
+1. supertypes: `kotlin/coroutines/jvm/internal/ContinuationImpl`
+2. package-local label field of int. Package-local, considering the function uses it, and the function is outside of the class.
+3. package-local fields for spilled variables. Same.
+4. public final method `invokeSuspend` of type `(Ljava/lang/Object;)Ljava/lang/Object;`. It overrides `BaseContinuationImpl`'s
+`invokeSuspend`, which calls the function.
+5. public or package-private constructor: `<init>` of type `(Lkotlin/coroutines/Continuation;)V`, which calls
+`BaseContinuatonImpl`.
+
+### Local Suspend Functions
+
+Local functions are weird: the way the compiler generates them is different in back-ends. Local suspend functions are even stranger.
+
+#### Old JVM: Unerased Suspend Lambdas
+
+Old JVM back-end generates local functions as lambdas. Thus, suspend local functions are generated as suspend lambda:
+```kotlin
+fun main() {
+    suspend fun local(i: Int) {}
+}
+```
+is generated as something like:
+```kotlin
+fun main() {
+    val local: suspend (Int) -> Unit = {}
+}
+```
+Doing so allows us to reuse the logic of captured variables and simplify the logic of code generation. However, because of the limitations
+of old BE, its `create` and `invoke` are unerased. In other words, the compiler duplicates them, generating unerased and erased copies.
+Unerased copy accepts typed parameters, and it contains the logic of lambda's `invoke` or `create`. The other accepts only `Any?`
+parameters since they override supertype's functions, and delegates call to an unerased copy. Thus, the layout of local suspend functions is:
+1. supertypes: `kotlin/coroutines/jvm/internal/SuspendLambda` and `kotlin/jvm/functions/Function{N}`
```

Comment:
[MD032 - Lists should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/main/doc/Rules.md#md032---lists-should-be-surrounded-by-blank-lines) here and below.